[LoopVectorize] Perform loop versioning for some early exit loops

david-arm opened the pull request:
When attempting to vectorise a loop with an uncountable early
exit, we attempt to discover if all the loads in the loop are
known to be dereferenceable. If at least one load could
potentially fault then we abandon vectorisation. This patch
adds support for vectorising loops with one potentially
faulting load by versioning the loop based on the load
pointer alignment. It is required that the vector load must
always fault on the first lane, i.e. the load should not
straddle a page boundary. Doing so ensures that the behaviour
of the vector and scalar loops is identical, i.e. if a load
does fault it will fault at the same scalar iteration.
Such vectorisation depends on the following conditions being
met:

1. The max vector width must not exceed the minimum page size.
This is done by adding a getMaxSafeVectorWidthInBits
wrapper that checks if we have an uncountable early exit.
For scalable vectors we must be able to determine the maximum
possible value of vscale.
2. The size of the loaded type must be a power of 2. This is
checked during legalisation.
3. The VF must be a power of two (so that the vector width can
divide wholly into the page size which is also power of 2).
For fixed-width vectors this is always true, and for scalable
vectors we query the TTI hook isVScaleKnownToBeAPowerOfTwo.
If the effective runtime VF could change during the loop then
this cannot be vectorised via loop versioning.
4. The load pointer must be aligned to a multiple of the vector
width. (NOTE: interleaving is currently disabled for these early
exit loops.) We add a runtime check to ensure this is true.

david-arm commented:
Gentle ping. :)

fhahn reviewed (commented):
Are there any documented guarantees related to alignment of pointers and dereferenceability of loads on the LLVM IR level? I had brief look in LangRef but couldn't find anything. 

Could the assumption cause problems with HW/software extensions, .e.g. something like AArch64 MTE or a software layer that checks out-of-bounds accesses? Not super familiar with the details of MTE, but I think it tracks sizes of allocations on smaller granularity than page sizes and traps on out-of-bound accesses.

david-arm commented:
> Are there any documented guarantees related to alignment of pointers and dereferenceability of loads on the LLVM IR level? I had brief look in LangRef but couldn't find anything.
> 
> Could the assumption cause problems with HW/software extensions, .e.g. something like AArch64 MTE or a software layer that checks out-of-bounds accesses? Not super familiar with the details of MTE, but I think it tracks sizes of allocations on smaller granularity than page sizes and traps on out-of-bound accesses.

This is something I'll look into - perhaps the worst case would be that loop versioning is off by default for AArch64 and would have to be enabled explicitly with a flag. Without digging into this in huge detail, is it possible that if MTE is a problem for this PR, then we also have a problem today in the loop vectoriser when we if-convert loads into unconditional loads based upon the result of `isDereferenceableAndAlignedInLoop` - see `LoopVectorizationLegality::canVectorizeWithIfConvert`. MTE can be used for stack tagging.

david-arm commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -10542,8 +10611,19 @@ bool LoopVectorizePass::processLoop(Loop *L) {
-    if (VF.Width.isVector() || SelectedIC > 1)
+    if (VF.Width.isVector() || SelectedIC > 1) {
+      if (LVL.getNumPotentiallyFaultingLoads()) {
+        assert(SelectedIC == 1 &&
+               "Interleaving not supported for early exit loops and "
+               "potentially faulting loads");
```
Yeah potentially, although currently we won't vectorise with UF>1 anyway because it's explicitly disabled in `selectInterleaveCount`. That's because we don't yet handle or'ing together the multiple early exit conditions for all the parts in vplan. So I didn't try adding support for runtime checks when UF>1 because there isn't any way to test it.

david-arm replied:
I've updated the code to handle interleaving, although we can't currently test it when IC > 1. However, it seems trivial/obvious enough that it should work.

david-arm commented on llvm/lib/Transforms/Vectorize/LoopVectorizationLegality.cpp:
```diff
@@ -1609,6 +1609,43 @@ bool LoopVectorizationLegality::canVectorizeLoopNestCFG(
+      return false;
+
+    LLVM_DEBUG(dbgs() << "LV: SCEV for Load Ptr: " << *Start << '\n');
+    PotentiallyFaultingLoads.push_back({LI, Start});
```
We need the Start because that's guaranteed to be loop-invariant, whereas `LI->getPointerOperand()` will be a loop-varying expression, i.e. `getelementptr i8, ptr %p, i64 %index` where `%index` comes from a PHI, or something like that. We have to add the pointer checks in the loop preheader.

In addition to `Start`, we also need to know the pointer type and the loaded type, which is why I thought it was easier to pass in `LI`. However, I could just collect 3 things instead - `{Start, PtrType, MemType}`. This would be more general than restricting to loads only?

david-arm replied:
I've changed the list to now only require `Start` and the load type.

david-arm commented on llvm/lib/Transforms/Vectorize/LoopVectorizationLegality.cpp:
```diff
@@ -1609,6 +1609,43 @@ bool LoopVectorizationLegality::canVectorizeLoopNestCFG(
+  for (LoadInst *LI : *Loads) {
+    LLVM_DEBUG(dbgs() << "LV:   Load: " << *LI << '\n');
+    Value *Ptr = LI->getPointerOperand();
+    if (!Ptr)
+      return false;
```
Damn, forgot to look at this. I'm not sure to be honest. I thought I'd seen similar checks elsewhere, but I'll take another look.

fhahn replied:
It should always return a valid value.

david-arm commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -1585,6 +1591,22 @@ class LoopVectorizationCostModel {
+    return Legal->isSafeForAnyVectorWidth() &&
+           (!Legal->hasUncountableEarlyExit() ||
+            !Legal->getNumPotentiallyFaultingLoads());
+  }
```
Done

david-arm commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -2133,6 +2155,41 @@ class GeneratedRTChecks {
+    return F.getFnAttribute(Attribute::VScaleRange).getVScaleRangeMax();
+
+  return std::nullopt;
+}
```
Yeah, I think in an early version of the patch I required getMaxVScale for something, but never moved it. Anyway, it's back in it's original place now!

david-arm commented on llvm/lib/Transforms/Vectorize/LoopVectorizationLegality.cpp:
```diff
@@ -1735,15 +1772,18 @@ bool LoopVectorizationLegality::isVectorizableEarlyExitLoop() {
+  if (!isDereferenceableReadOnlyLoop(TheLoop, PSE.getSE(), DT, AC, &Predicates,
+                                     &Loads)) {
+    if (!TTI->getMinPageSize() || !analyzePotentiallyFaultingLoads(&Loads)) {
+      reportVectorizationFailure(
```
I've just removed `isDereferenceableReadOnlyLoop` now and inlined some of the logic into the loop above.

david-arm commented on llvm/lib/Analysis/Loads.cpp:
```diff
@@ -816,15 +816,26 @@ bool llvm::canReplacePointersIfEqual(const Value *From, const Value *To,
           return false;
-      } else if (I.mayReadFromMemory() || I.mayWriteToMemory() || I.mayThrow())
-        return false;
+        Result = false;
```
Done

david-arm commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -1585,6 +1591,22 @@ class LoopVectorizationCostModel {
+      MaxSafeVectorWidth =
+          std::min(MaxSafeVectorWidth, uint64_t(*TTI.getMinPageSize()) * 8);
+    return MaxSafeVectorWidth;
+  }
```
Done

david-arm reviewed (commented):
For now, I've changed it so that loop versioning is turned off by default. When the user is confident that MTE compatibility is not required they can enable it manually. In future, it would be good to revisit this because even with MTE we can still vectorise with max vector widths of 16 bytes, and memory tagging is currently only supported in LLVM for Android targets anyway.

fhahn commented on llvm/include/llvm/Transforms/Vectorize/LoopVectorizationLegality.h:
```diff
@@ -423,6 +430,19 @@ class LoopVectorizationLegality {
+  /// uncountable early exits.
+  const SmallVectorImpl<std::pair<const SCEV *, Type *>> *
+  getPotentiallyFaultingPointers() const {
+    return &PotentiallyFaultingPtrs;
```
Can we just return an `ArrayRef` here, or do we need to return a pointer

david-arm replied:
Done

fhahn commented on llvm/include/llvm/Transforms/Vectorize/LoopVectorizationLegality.h:
```diff
@@ -528,6 +548,11 @@ class LoopVectorizationLegality {
+  /// Returns true if all loads in the loop contained in \p Loads can be
+  /// analyzed as potentially faulting. Any loads that may fault are added to
+  /// the member variable PotentiallyFaultingPtrs.
+  bool analyzePotentiallyFaultingLoads(SmallVectorImpl<LoadInst *> *Loads);
```
Can pass by reference?

david-arm replied:
Done

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -2163,6 +2169,27 @@ class GeneratedRTChecks {
+        SE->getMulExpr(ScevEC, SE->getConstant(EltSize),
+                       (SCEV::NoWrapFlags)(SCEV::FlagNSW | SCEV::FlagNUW));
+    const SCEV *Rem = SE->getURemExpr(Start, Align);
+    PSE.addPredicate(*(SE->getEqualPredicate(Rem, SE->getZero(PtrIntType))));
```
```suggestion
    PSE.addPredicate(*SE->getEqualPredicate(Rem, SE->getZero(PtrIntType)));
```

david-arm replied:
Done

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -2163,6 +2169,27 @@ class GeneratedRTChecks {
 } // namespace
 
+static void addPointerAlignmentChecks(
+    const SmallVectorImpl<std::pair<const SCEV *, Type *>> *Ptrs, Function *F,
```
```suggestion
    ArrayRefl<std::pair<const SCEV *, Type *> Ptrs, Function *F,
```

david-arm replied:
Done

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorizationLegality.cpp:
```diff
@@ -1609,6 +1609,43 @@ bool LoopVectorizationLegality::canVectorizeLoopNestCFG(
+    const SCEV *PtrExpr = PSE.getSCEV(Ptr);
+    const SCEVAddRecExpr *AR = dyn_cast<SCEVAddRecExpr>(PtrExpr);
+    // TODO: Deal with loop invariant pointers.
+    if (!AR || AR->getLoop() != TheLoop || !AR->isAffine())
```
Would be good to explain here that we check for add-recs because the reasoning is only safe if the load executes at least once?

We should probably also restrict this just for loads from the default address space (0) for now

david-arm replied:
Done

fhahn reviewed (commented):
> This is something I'll look into - perhaps the worst case for now would be that loop versioning is off by default for AArch64 and would have to be enabled explicitly with a flag. Without digging into this in huge detail, is it possible that if MTE is a problem for this PR, then we also have a problem today in the loop vectoriser when we if-convert loads into unconditional loads based upon the result of `isDereferenceableAndAlignedInLoop` - see `LoopVectorizationLegality::canVectorizeWithIfConvert`. MTE can be used for stack tagging.

IIRC `isDereferenceableAndAlignedInLoop` determines dereferenceability separately from either attributes or known allocation sizes, so it should never return true if we could access out-of-bounds I think. I am not sure how the current uses would cause problems with extensions like MTE, is there any particular issue you are thinking about?

nikic reviewed (changes requested):
Do I understand correctly that this is generating out-of-bounds loads and relying on page granularity to guarantee they don't trap? If so, we cannot perform this transform with normal loads, because it is UB at the IR level -- the behavior of the underlying hardware is irrelevant.

You're going to need a new load intrinsic to support this. I have this RFC draft on the topic lying around: https://hackmd.io/@nikic/S1O4QWYZkx I haven't submitted it, because I'm not particularly happy with the `%defined_size` parameter, which is needed to specify the operational semantics of the intrinsic, but not relevant for lowering.

programmerjake commented:
I proposed a similar speculative load instruction here: https://discourse.llvm.org/t/rfc-supporting-more-early-exit-loops/84690/2?u=programmerjake

david-arm commented:
> For RVV first faulting loads, a VP variant intrinsic could support this, for example `declare { <8 x i8>, i32 } @llvm.vp.load.ff.v8i8.p0(ptr %ptr, <8 x i1> %mask, i32 %evl)`. It differs by returning a structure that includes both the loaded data and latest EVL.

This patch is explicitly avoiding using first faulting loads, by relying upon a combination of vector load alignment and knowledge of the hardware minimum page size. I believe @huntergr-arm recently created a RFC about an alternative approach using first faulting loads to vectorise more complex loops involving multiple exits and/or stores in the loop.

If I understand correctly, the point raised by @nikic is that the semantics of the IR load operation mean that if any portion of the vector load is out-of-bounds (even though we know it won't fault) then the entire load is UB. This essentially means the entire vectorised loop becomes UB and so in theory could be optimised away, even if in practice for existing tests and benchmark it works. Using a new out-of-bounds load intrinsic with different semantics avoids this problem, and still allows us to achieve the best performance.

david-arm commented:
Hi @nikic, just for my own understanding can you point me to the part of the LangRef that says out-of-bounds loads are UB? Perhaps I'm missing something, but I did have a look and it wasn't immediately obvious to me where this is stated.

fhahn commented:
> Hi @nikic, just for my own understanding can you point me to the part of the LangRef that says out-of-bounds loads are UB? Perhaps I'm missing something, but I did have a look and it wasn't immediately obvious to me where this is stated.

This was my understanding as well, but I also couldn't find a clear reference from LangRef. Put up https://github.com/llvm/llvm-project/pull/128429 to clarify

david-arm committed:
Add tests

david-arm committed:
[LoopVectorize] Perform loop versioning for some early exit loops

When attempting to vectorise a loop with an uncountable early
exit, we attempt to discover if all the loads in the loop are
known to be dereferenceable. If at least one load could
potentially fault then we abandon vectorisation. This patch
adds support for vectorising loops with one potentially
faulting load by versioning the loop based on the load
pointer alignment. It is required that the vector load must
always fault on the first lane, i.e. the load should not
straddle a page boundary. Doing so ensures that the behaviour
of the vector and scalar loops is identical, i.e. if a load
does fault it will fault at the same scalar iteration.
Such vectorisation depends on the following conditions being
met:

1. The max vector width must not exceed the minimum page size.
This is done by adding a getMaxSafeVectorWidthInBits
wrapper that checks if we have an uncountable early exit.
For scalable vectors we must be able to determine the maximum
possible value of vscale.
2. The size of the loaded type must be a power of 2. This is
checked during legalisation.
3. The VF must be a power of two (so that the vector width can
divide wholly into the page size which is also power of 2).
For fixed-width vectors this is always true, and for scalable
vectors we query the TTI hook isVScaleKnownToBeAPowerOfTwo.
If the effective runtime VF could change during the loop then
this cannot be vectorised via loop versioning.
4. The load pointer must be aligned to a multiple of the vector
width. (NOTE: interleaving is currently disabled for these early
exit loops.) We add a runtime check to ensure this is true.

david-arm committed:
Address review comments

david-arm reviewed (commented):
Updated the patch to deal with review comments and also fix a bug I spotted where I treated loads as being unconditionally dereferenceable after the early exit - see the new test `same_exit_block_no_live_outs_faulting_load_after_early_exit`. This is not true because the first lane of the vector load could be faulting, which would not have happened with the scalar loop.

david-arm committed:
Fix formatting issues
